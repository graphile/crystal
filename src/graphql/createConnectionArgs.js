import { memoize, fromPairs, includes, snakeCase, toUpper } from 'lodash'
import { GraphQLEnumType, GraphQLInt, GraphQLBoolean } from 'graphql'
import { CursorType } from './types.js'
import getType from './getType.js'

const createConnectionArgs = (table, ignoreColumnConditions = []) => ({
  // The column specified by `orderBy` means more than just the order to
  // return items in. This column is also the column we will use for
  // cursors.
  orderBy: {
    type: createTableOrderingEnum(table),
    description:
      'The order the resulting items should be returned in. This argument ' +
      'is also important as it is used in creating pagination cursors. This ' +
      'value’s default is the primary key for the object.',
    defaultValue: (() => {
      const column = table.getPrimaryKeys()[0]
      if (column) return column.name
      return null
    })(),
  },
  first: {
    type: GraphQLInt,
    description:
      'The top `n` items in the set to be returned. Can’t be used ' +
      'with `last`.',
  },
  last: {
    type: GraphQLInt,
    description:
      'The bottom `n` items in the set to be returned. Can’t be used ' +
      'with `first`.',
  },
  before: {
    type: CursorType,
    description: 'Constrains the set to nodes *before* this cursor in the specified ordering.',
  },
  after: {
    type: CursorType,
    description: 'Constrains the set to nodes *after* this cursor in the specified ordering.',
  },
  offset: {
    type: GraphQLInt,
    description: 'An integer offset representing how many items to skip in the set.',
  },
  descending: {
    type: GraphQLBoolean,
    description:
      'If `true` the nodes will be in descending order, if `false` the ' +
      'items will be in ascending order. `false` by default.',
    defaultValue: false,
  },
  ...fromPairs(
    table
    .getColumns()
    // If `ignoreColumnConditions` is set to true, all column conditions will
    // be disabled. If `ignoreColumnConditions` is an array, only certain
    // conditions will be ignored.
    .filter(column => (ignoreColumnConditions === true ? false : !includes(ignoreColumnConditions, column)))
    .map(column => [column.getFieldName(), {
      type: getType(column.type),
      description:
        'Filters the resulting set with an equality test on the ' +
        `${column.getMarkdownFieldName()} field.`,
      // TODO: Deprecate this…
      // deprecationReason:
      //   'Simple equality testing is insufficient for the nodes field and just ' +
      //   'adds noise. Instead use procedures for custom set filtering.',
    }])
  ),
})

export default createConnectionArgs

/**
 * Creates an ordering enum which simply contains all of a `Table`s columns.
 *
 * @param {Table} table
 * @returns {GraphQLEnumType}
 */
// TODO: Some way to eliminate some columns from ordering enum?
export const createTableOrderingEnum = memoize(table =>
  new GraphQLEnumType({
    name: `${table.getTypeName()}Ordering`,
    description: `Properties with which ${table.getMarkdownTypeName()} can be ordered.`,

    values: fromPairs(
      table
      .getColumns()
      .map(column => [toUpper(snakeCase(column.getFieldName())), {
        value: column.name,
        description: column.description,
      }])
    ),
  }))
