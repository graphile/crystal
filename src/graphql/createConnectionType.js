import { memoize } from 'lodash'
import createTableType from './createTableType.js'
import { PageInfoType, CursorType } from './types.js'

import {
  GraphQLObjectType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLInt,
} from 'graphql'

const createTableConnectionType = memoize(table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Connection`,
    description: `A connection to a list of ${table.getMarkdownTypeName()} items`,

    // TODO: Implement a `ConnectionType` interface

    fields: {
      pageInfo: {
        type: new GraphQLNonNull(PageInfoType),
        description: `Information to aid in pagination of type ${table.getMarkdownTypeName()}.`,
        resolve: pageInfo => pageInfo,
      },
      totalCount: {
        type: GraphQLInt,
        description: 'All of the items available to be queried in this connection.',
        resolve: ({ totalCount }) => totalCount,
      },
      nodes: {
        type: new GraphQLList(createTableType(table)),
        description: `The queried list of ${table.getMarkdownTypeName()}.`,
        resolve: ({ nodes }) => nodes,
      },
      edges: {
        type: new GraphQLList(createTableEdgeType(table)),
        description: 'A single item and a cursor to aid in pagination.',
        resolve: ({ edges }) => edges,
      },
    },
  })
)

export default createTableConnectionType

const createTableEdgeType = table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Edge`,
    description: `An edge in the \`${table.getTypeName()}Connection\`.`,

    fields: {
      cursor: {
        type: new GraphQLNonNull(CursorType),
        description: 'The cursor describing the position of the edgeâ€™s associated node.',
        resolve: ({ cursor }) => cursor || 'null',
      },
      node: {
        type: createTableType(table),
        description: 'The item at the end of the edge.',
        resolve: ({ node }) => node,
      },
    },
  })
