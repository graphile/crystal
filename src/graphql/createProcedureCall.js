const createProcedureCall = procedure => {
  // Construct the qualified procedure name.
  const procedureName = `"${procedure.schema.name}"."${procedure.name}"`

  // Construct the argument list for the procedure call.
  const procedureArgsList = Array.from(procedure.args).map(([name, type], i) => {
    const placeholder = `$${i + 1}`

    // If the type of this argument is a table type, we will be expecting JSON
    // but we need that JSON to be a table. Therefore we run it through
    // `json_populate_record`.
    if (type.isTableType)
      return `json_populate_record(null::"${type.table.schema.name}"."${type.table.name}", ${placeholder})`

    return placeholder
  }).join(', ')

  // Add the procedure name with the procedure argument list.
  const procedureCall = `${procedureName}(${procedureArgsList})`

  return procedureCall
}

export default createProcedureCall
