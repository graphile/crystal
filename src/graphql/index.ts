import createGraphQLSchema from './schema/createGQLSchema'

export {
  createGraphQLSchema,
}
