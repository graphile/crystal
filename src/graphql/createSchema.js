import { GraphQLSchema } from 'graphql'
import createQueryType from './query/createQueryType.js'
import createMutationType from './mutation/createMutationType.js'

/**
 * Creates a GraphQLSchema from a PostgreSQL schema.
 *
 * @param {Schema} schema
 * @returns {GrpahQLSchema}
 */
const createGraphqlSchema = (schema, options) =>
  new GraphQLSchema({
    query: createQueryType(schema, options),
    mutation: createMutationType(schema, options),
  })

export default createGraphqlSchema
