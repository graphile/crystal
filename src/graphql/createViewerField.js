import { memoize } from 'lodash'
import { GraphQLObjectType, GraphQLNonNull } from 'graphql'
import createQueryFields from './query/createQueryFields.js'

const createViewerField = memoize(schema => ({
  type: new GraphQLNonNull(new GraphQLObjectType({
    name: 'Viewer',
    description: '',
    fields: createQueryFields(schema),
  })),

  description:
    'A single entry query for advanced data clients like Relay. Nothing ' +
    'special at all, if you don’t know what this field is for, you probably ' +
    'don’t need it.',

  resolve: source => source || {},
}))

export default createViewerField
