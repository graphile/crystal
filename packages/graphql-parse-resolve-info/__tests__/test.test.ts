import {
  parseResolveInfo,
  simplifyParsedResolveInfoFragmentWithType,
} from "../src";
import {
  graphql,
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLScalarType,
  GraphQLSchema,
} from "graphql";
import { Kind } from "graphql/language";
import assert from "assert";

const query = `
  query Test($include: Boolean!, $exclude: Boolean!) {
    allPosts {
      edges {
        cursor @include(if: $include)
        node {
          ...PostDetails
          author: personByAuthorId {
            firstPost {
              ...PostDetails
            }
            friends {
              nodes {
                ...PersonDetails
              }
              totalCount @include(if: $include)
              pageInfo @skip(if: $exclude) {
                startCursor
              }
            }
          }
        }
      }
    }
  }

  fragment PersonDetails on Person {
    id
    ...MorePersonDetails @include(if: $include)
    lastName @include(if: false)
    bio @skip(if: true)
  }

  fragment MorePersonDetails on Person {
    name
    firstName
  }

  fragment PostDetails on Post {
    id
    headline
    headlineTrimmed @skip(if: $exclude)
    author: personByAuthorId {
      ...PersonDetails
    }
  }
`;

const Cursor = new GraphQLScalarType({
  name: "Cursor",
  serialize: (value) => String(value),
  parseValue: (value) => String(value),
  parseLiteral: (ast) => {
    if (ast.kind !== Kind.STRING) {
      throw new Error("Can only parse string values");
    }
    return ast.value;
  },
});

const PageInfo = new GraphQLObjectType({
  name: "PageInfo",
  fields: {
    startCursor: {
      type: Cursor,
    },
  },
});

const Person: GraphQLObjectType = new GraphQLObjectType({
  name: "Person",
  fields: () => ({
    id: {
      type: GraphQLString,
    },
    name: {
      type: GraphQLString,
    },
    firstName: {
      type: GraphQLString,
    },
    lastName: {
      type: GraphQLString,
    },
    bio: {
      type: GraphQLString,
    },
    firstPost: {
      type: Post,
    },
    friends: {
      type: PersonConnection,
    },
  }),
});

const PersonEdge = new GraphQLObjectType({
  name: "PersonEdge",
  fields: {
    cursor: {
      type: Cursor,
    },
    node: {
      type: Person,
    },
  },
});

const PersonConnection = new GraphQLObjectType({
  name: "PersonConnection",
  fields: {
    edges: {
      type: new GraphQLList(PersonEdge),
    },
    nodes: {
      type: new GraphQLList(Person),
    },
    pageInfo: {
      type: PageInfo,
    },
    totalCount: {
      type: GraphQLInt,
    },
  },
});

const Post = new GraphQLObjectType({
  name: "Post",
  fields: {
    personByAuthorId: {
      type: Person,
    },
    id: {
      type: GraphQLString,
    },
    headline: {
      type: GraphQLString,
    },
    headlineTrimmed: {
      type: GraphQLString,
    },
  },
});

const PostEdge = new GraphQLObjectType({
  name: "PostEdge",
  fields: {
    cursor: {
      type: Cursor,
    },
    node: {
      type: Post,
    },
  },
});

const PostsConnection = new GraphQLObjectType({
  name: "PostsConnection",
  fields: {
    edges: {
      type: new GraphQLList(PostEdge),
    },
  },
});

const Query = new GraphQLObjectType({
  name: "Query",
  fields: {
    allPosts: {
      type: new GraphQLNonNull(PostsConnection),
      resolve(parent, args, context, resolveInfo) {
        const parsedResolveInfoFragment = parseResolveInfo(resolveInfo);
        assert(parsedResolveInfoFragment);
        const simplifiedFragment = simplifyParsedResolveInfoFragmentWithType(
          parsedResolveInfoFragment,
          resolveInfo.returnType,
        );
        context.test({
          parsedResolveInfoFragment,
          simplifiedFragment,
        });
        return [];
        // ...
      },
    },

    // ...
  },
});

const Schema = new GraphQLSchema({
  query: Query,
});

test("basic", async () => {
  const variables = {
    include: true,
    exclude: false,
  };
  const { parsedResolveInfoFragment, simplifiedFragment } = await new Promise(
    (resolve, reject) => {
      let o: any;
      graphql(
        Schema,
        query,
        null,
        {
          test: (_o: any) => (o = _o),
        },
        variables,
      ).then((d) => {
        try {
          const { errors } = d;
          expect(errors).toBeFalsy();
        } catch (e) {
          return reject(e);
        }
        if (o) {
          resolve(o);
        } else {
          reject(new Error("test not called?"));
        }
      }, reject);
    },
  );
  expect(parsedResolveInfoFragment).toMatchSnapshot();
  expect(simplifiedFragment).toMatchSnapshot();
});

test("directives", async () => {
  const variables = {
    include: false,
    exclude: true,
  };
  const { parsedResolveInfoFragment, simplifiedFragment } = await new Promise(
    (resolve, reject) => {
      let o: any;
      graphql(
        Schema,
        query,
        null,
        {
          test: (_o: any) => (o = _o),
        },
        variables,
      ).then((d) => {
        try {
          const { errors } = d;
          expect(errors).toBeFalsy();
        } catch (e) {
          return reject(e);
        }
        if (o) {
          resolve(o);
        } else {
          reject(new Error("test not called?"));
        }
      }, reject);
    },
  );
  expect(parsedResolveInfoFragment).toMatchSnapshot();
  expect(simplifiedFragment).toMatchSnapshot();
});
