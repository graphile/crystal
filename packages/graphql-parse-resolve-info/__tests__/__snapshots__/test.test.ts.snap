// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Object {
  "alias": "allPosts",
  "args": Object {},
  "fieldsByTypeName": Object {
    "PostsConnection": Object {
      "edges": Object {
        "alias": "edges",
        "args": Object {},
        "fieldsByTypeName": Object {
          "PostEdge": Object {
            "cursor": Object {
              "alias": "cursor",
              "args": Object {},
              "fieldsByTypeName": Object {},
              "name": "cursor",
            },
            "node": Object {
              "alias": "node",
              "args": Object {},
              "fieldsByTypeName": Object {
                "Post": Object {
                  "author": Object {
                    "alias": "author",
                    "args": Object {},
                    "fieldsByTypeName": Object {
                      "Person": Object {
                        "firstName": Object {
                          "alias": "firstName",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "firstName",
                        },
                        "firstPost": Object {
                          "alias": "firstPost",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "Post": Object {
                              "author": Object {
                                "alias": "author",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "firstName": Object {
                                      "alias": "firstName",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "firstName",
                                    },
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                    "name": Object {
                                      "alias": "name",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "name",
                                    },
                                  },
                                },
                                "name": "personByAuthorId",
                              },
                              "headline": Object {
                                "alias": "headline",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headline",
                              },
                              "headlineTrimmed": Object {
                                "alias": "headlineTrimmed",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headlineTrimmed",
                              },
                              "id": Object {
                                "alias": "id",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "id",
                              },
                            },
                          },
                          "name": "firstPost",
                        },
                        "friends": Object {
                          "alias": "friends",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "PersonConnection": Object {
                              "nodes": Object {
                                "alias": "nodes",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "firstName": Object {
                                      "alias": "firstName",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "firstName",
                                    },
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                    "name": Object {
                                      "alias": "name",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "name",
                                    },
                                  },
                                },
                                "name": "nodes",
                              },
                              "pageInfo": Object {
                                "alias": "pageInfo",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "PageInfo": Object {
                                    "startCursor": Object {
                                      "alias": "startCursor",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "startCursor",
                                    },
                                  },
                                },
                                "name": "pageInfo",
                              },
                              "totalCount": Object {
                                "alias": "totalCount",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "totalCount",
                              },
                            },
                          },
                          "name": "friends",
                        },
                        "id": Object {
                          "alias": "id",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "id",
                        },
                        "name": Object {
                          "alias": "name",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "name",
                        },
                      },
                    },
                    "name": "personByAuthorId",
                  },
                  "headline": Object {
                    "alias": "headline",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headline",
                  },
                  "headlineTrimmed": Object {
                    "alias": "headlineTrimmed",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headlineTrimmed",
                  },
                  "id": Object {
                    "alias": "id",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "id",
                  },
                },
              },
              "name": "node",
            },
          },
        },
        "name": "edges",
      },
    },
  },
  "name": "allPosts",
}
`;

exports[`basic 2`] = `
Object {
  "alias": "allPosts",
  "args": Object {},
  "fields": Object {
    "edges": Object {
      "alias": "edges",
      "args": Object {},
      "fieldsByTypeName": Object {
        "PostEdge": Object {
          "cursor": Object {
            "alias": "cursor",
            "args": Object {},
            "fieldsByTypeName": Object {},
            "name": "cursor",
          },
          "node": Object {
            "alias": "node",
            "args": Object {},
            "fieldsByTypeName": Object {
              "Post": Object {
                "author": Object {
                  "alias": "author",
                  "args": Object {},
                  "fieldsByTypeName": Object {
                    "Person": Object {
                      "firstName": Object {
                        "alias": "firstName",
                        "args": Object {},
                        "fieldsByTypeName": Object {},
                        "name": "firstName",
                      },
                      "firstPost": Object {
                        "alias": "firstPost",
                        "args": Object {},
                        "fieldsByTypeName": Object {
                          "Post": Object {
                            "author": Object {
                              "alias": "author",
                              "args": Object {},
                              "fieldsByTypeName": Object {
                                "Person": Object {
                                  "firstName": Object {
                                    "alias": "firstName",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "firstName",
                                  },
                                  "id": Object {
                                    "alias": "id",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "id",
                                  },
                                  "name": Object {
                                    "alias": "name",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "name",
                                  },
                                },
                              },
                              "name": "personByAuthorId",
                            },
                            "headline": Object {
                              "alias": "headline",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "headline",
                            },
                            "headlineTrimmed": Object {
                              "alias": "headlineTrimmed",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "headlineTrimmed",
                            },
                            "id": Object {
                              "alias": "id",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "id",
                            },
                          },
                        },
                        "name": "firstPost",
                      },
                      "friends": Object {
                        "alias": "friends",
                        "args": Object {},
                        "fieldsByTypeName": Object {
                          "PersonConnection": Object {
                            "nodes": Object {
                              "alias": "nodes",
                              "args": Object {},
                              "fieldsByTypeName": Object {
                                "Person": Object {
                                  "firstName": Object {
                                    "alias": "firstName",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "firstName",
                                  },
                                  "id": Object {
                                    "alias": "id",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "id",
                                  },
                                  "name": Object {
                                    "alias": "name",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "name",
                                  },
                                },
                              },
                              "name": "nodes",
                            },
                            "pageInfo": Object {
                              "alias": "pageInfo",
                              "args": Object {},
                              "fieldsByTypeName": Object {
                                "PageInfo": Object {
                                  "startCursor": Object {
                                    "alias": "startCursor",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "startCursor",
                                  },
                                },
                              },
                              "name": "pageInfo",
                            },
                            "totalCount": Object {
                              "alias": "totalCount",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "totalCount",
                            },
                          },
                        },
                        "name": "friends",
                      },
                      "id": Object {
                        "alias": "id",
                        "args": Object {},
                        "fieldsByTypeName": Object {},
                        "name": "id",
                      },
                      "name": Object {
                        "alias": "name",
                        "args": Object {},
                        "fieldsByTypeName": Object {},
                        "name": "name",
                      },
                    },
                  },
                  "name": "personByAuthorId",
                },
                "headline": Object {
                  "alias": "headline",
                  "args": Object {},
                  "fieldsByTypeName": Object {},
                  "name": "headline",
                },
                "headlineTrimmed": Object {
                  "alias": "headlineTrimmed",
                  "args": Object {},
                  "fieldsByTypeName": Object {},
                  "name": "headlineTrimmed",
                },
                "id": Object {
                  "alias": "id",
                  "args": Object {},
                  "fieldsByTypeName": Object {},
                  "name": "id",
                },
              },
            },
            "name": "node",
          },
        },
      },
      "name": "edges",
    },
  },
  "fieldsByTypeName": Object {
    "PostsConnection": Object {
      "edges": Object {
        "alias": "edges",
        "args": Object {},
        "fieldsByTypeName": Object {
          "PostEdge": Object {
            "cursor": Object {
              "alias": "cursor",
              "args": Object {},
              "fieldsByTypeName": Object {},
              "name": "cursor",
            },
            "node": Object {
              "alias": "node",
              "args": Object {},
              "fieldsByTypeName": Object {
                "Post": Object {
                  "author": Object {
                    "alias": "author",
                    "args": Object {},
                    "fieldsByTypeName": Object {
                      "Person": Object {
                        "firstName": Object {
                          "alias": "firstName",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "firstName",
                        },
                        "firstPost": Object {
                          "alias": "firstPost",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "Post": Object {
                              "author": Object {
                                "alias": "author",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "firstName": Object {
                                      "alias": "firstName",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "firstName",
                                    },
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                    "name": Object {
                                      "alias": "name",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "name",
                                    },
                                  },
                                },
                                "name": "personByAuthorId",
                              },
                              "headline": Object {
                                "alias": "headline",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headline",
                              },
                              "headlineTrimmed": Object {
                                "alias": "headlineTrimmed",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headlineTrimmed",
                              },
                              "id": Object {
                                "alias": "id",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "id",
                              },
                            },
                          },
                          "name": "firstPost",
                        },
                        "friends": Object {
                          "alias": "friends",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "PersonConnection": Object {
                              "nodes": Object {
                                "alias": "nodes",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "firstName": Object {
                                      "alias": "firstName",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "firstName",
                                    },
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                    "name": Object {
                                      "alias": "name",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "name",
                                    },
                                  },
                                },
                                "name": "nodes",
                              },
                              "pageInfo": Object {
                                "alias": "pageInfo",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "PageInfo": Object {
                                    "startCursor": Object {
                                      "alias": "startCursor",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "startCursor",
                                    },
                                  },
                                },
                                "name": "pageInfo",
                              },
                              "totalCount": Object {
                                "alias": "totalCount",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "totalCount",
                              },
                            },
                          },
                          "name": "friends",
                        },
                        "id": Object {
                          "alias": "id",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "id",
                        },
                        "name": Object {
                          "alias": "name",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "name",
                        },
                      },
                    },
                    "name": "personByAuthorId",
                  },
                  "headline": Object {
                    "alias": "headline",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headline",
                  },
                  "headlineTrimmed": Object {
                    "alias": "headlineTrimmed",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headlineTrimmed",
                  },
                  "id": Object {
                    "alias": "id",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "id",
                  },
                },
              },
              "name": "node",
            },
          },
        },
        "name": "edges",
      },
    },
  },
  "name": "allPosts",
}
`;

exports[`directives 1`] = `
Object {
  "alias": "allPosts",
  "args": Object {},
  "fieldsByTypeName": Object {
    "PostsConnection": Object {
      "edges": Object {
        "alias": "edges",
        "args": Object {},
        "fieldsByTypeName": Object {
          "PostEdge": Object {
            "node": Object {
              "alias": "node",
              "args": Object {},
              "fieldsByTypeName": Object {
                "Post": Object {
                  "author": Object {
                    "alias": "author",
                    "args": Object {},
                    "fieldsByTypeName": Object {
                      "Person": Object {
                        "firstPost": Object {
                          "alias": "firstPost",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "Post": Object {
                              "author": Object {
                                "alias": "author",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                  },
                                },
                                "name": "personByAuthorId",
                              },
                              "headline": Object {
                                "alias": "headline",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headline",
                              },
                              "id": Object {
                                "alias": "id",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "id",
                              },
                            },
                          },
                          "name": "firstPost",
                        },
                        "friends": Object {
                          "alias": "friends",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "PersonConnection": Object {
                              "nodes": Object {
                                "alias": "nodes",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                  },
                                },
                                "name": "nodes",
                              },
                            },
                          },
                          "name": "friends",
                        },
                        "id": Object {
                          "alias": "id",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "id",
                        },
                      },
                    },
                    "name": "personByAuthorId",
                  },
                  "headline": Object {
                    "alias": "headline",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headline",
                  },
                  "id": Object {
                    "alias": "id",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "id",
                  },
                },
              },
              "name": "node",
            },
          },
        },
        "name": "edges",
      },
    },
  },
  "name": "allPosts",
}
`;

exports[`directives 2`] = `
Object {
  "alias": "allPosts",
  "args": Object {},
  "fields": Object {
    "edges": Object {
      "alias": "edges",
      "args": Object {},
      "fieldsByTypeName": Object {
        "PostEdge": Object {
          "node": Object {
            "alias": "node",
            "args": Object {},
            "fieldsByTypeName": Object {
              "Post": Object {
                "author": Object {
                  "alias": "author",
                  "args": Object {},
                  "fieldsByTypeName": Object {
                    "Person": Object {
                      "firstPost": Object {
                        "alias": "firstPost",
                        "args": Object {},
                        "fieldsByTypeName": Object {
                          "Post": Object {
                            "author": Object {
                              "alias": "author",
                              "args": Object {},
                              "fieldsByTypeName": Object {
                                "Person": Object {
                                  "id": Object {
                                    "alias": "id",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "id",
                                  },
                                },
                              },
                              "name": "personByAuthorId",
                            },
                            "headline": Object {
                              "alias": "headline",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "headline",
                            },
                            "id": Object {
                              "alias": "id",
                              "args": Object {},
                              "fieldsByTypeName": Object {},
                              "name": "id",
                            },
                          },
                        },
                        "name": "firstPost",
                      },
                      "friends": Object {
                        "alias": "friends",
                        "args": Object {},
                        "fieldsByTypeName": Object {
                          "PersonConnection": Object {
                            "nodes": Object {
                              "alias": "nodes",
                              "args": Object {},
                              "fieldsByTypeName": Object {
                                "Person": Object {
                                  "id": Object {
                                    "alias": "id",
                                    "args": Object {},
                                    "fieldsByTypeName": Object {},
                                    "name": "id",
                                  },
                                },
                              },
                              "name": "nodes",
                            },
                          },
                        },
                        "name": "friends",
                      },
                      "id": Object {
                        "alias": "id",
                        "args": Object {},
                        "fieldsByTypeName": Object {},
                        "name": "id",
                      },
                    },
                  },
                  "name": "personByAuthorId",
                },
                "headline": Object {
                  "alias": "headline",
                  "args": Object {},
                  "fieldsByTypeName": Object {},
                  "name": "headline",
                },
                "id": Object {
                  "alias": "id",
                  "args": Object {},
                  "fieldsByTypeName": Object {},
                  "name": "id",
                },
              },
            },
            "name": "node",
          },
        },
      },
      "name": "edges",
    },
  },
  "fieldsByTypeName": Object {
    "PostsConnection": Object {
      "edges": Object {
        "alias": "edges",
        "args": Object {},
        "fieldsByTypeName": Object {
          "PostEdge": Object {
            "node": Object {
              "alias": "node",
              "args": Object {},
              "fieldsByTypeName": Object {
                "Post": Object {
                  "author": Object {
                    "alias": "author",
                    "args": Object {},
                    "fieldsByTypeName": Object {
                      "Person": Object {
                        "firstPost": Object {
                          "alias": "firstPost",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "Post": Object {
                              "author": Object {
                                "alias": "author",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                  },
                                },
                                "name": "personByAuthorId",
                              },
                              "headline": Object {
                                "alias": "headline",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "headline",
                              },
                              "id": Object {
                                "alias": "id",
                                "args": Object {},
                                "fieldsByTypeName": Object {},
                                "name": "id",
                              },
                            },
                          },
                          "name": "firstPost",
                        },
                        "friends": Object {
                          "alias": "friends",
                          "args": Object {},
                          "fieldsByTypeName": Object {
                            "PersonConnection": Object {
                              "nodes": Object {
                                "alias": "nodes",
                                "args": Object {},
                                "fieldsByTypeName": Object {
                                  "Person": Object {
                                    "id": Object {
                                      "alias": "id",
                                      "args": Object {},
                                      "fieldsByTypeName": Object {},
                                      "name": "id",
                                    },
                                  },
                                },
                                "name": "nodes",
                              },
                            },
                          },
                          "name": "friends",
                        },
                        "id": Object {
                          "alias": "id",
                          "args": Object {},
                          "fieldsByTypeName": Object {},
                          "name": "id",
                        },
                      },
                    },
                    "name": "personByAuthorId",
                  },
                  "headline": Object {
                    "alias": "headline",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "headline",
                  },
                  "id": Object {
                    "alias": "id",
                    "args": Object {},
                    "fieldsByTypeName": Object {},
                    "name": "id",
                  },
                },
              },
              "name": "node",
            },
          },
        },
        "name": "edges",
      },
    },
  },
  "name": "allPosts",
}
`;
